### PR TITLE
SHA-1 and -2 performance improvements

### PR DESCRIPTION
**Pull request by coruus**
See #322 for the original PR, #391 for the erroneous PR that got merged, and #402 and #404 for the PRs reverting . Test output for `closure/goog/crypt`:

```
00:38:26 : closure/goog/crypt/aes_test.html : Loading
00:38:26 : closure/goog/crypt/aes_test.html : Passed
00:38:26 : closure/goog/crypt/arc4_test.html : Loading
00:38:27 : closure/goog/crypt/arc4_test.html : Passed
00:38:27 : .../closure/goog/crypt/base64_test.html : Loading
00:38:28 : .../closure/goog/crypt/base64_test.html : Passed
00:38:28 : closure/goog/crypt/basen_test.html : Loading
00:38:28 : closure/goog/crypt/basen_test.html : Passed
00:38:28 : .../closure/goog/crypt/blobhasher_test.html : Loading
00:38:29 : .../closure/goog/crypt/blobhasher_test.html : Passed
00:38:29 : closure/goog/crypt/cbc_test.html : Loading
00:38:30 : closure/goog/crypt/cbc_test.html : Passed
00:38:30 : closure/goog/crypt/crypt_test.html : Loading
00:38:31 : closure/goog/crypt/crypt_test.html : Passed
00:38:31 : .../closure/goog/crypt/hash32_test.html : Loading
00:38:31 : .../closure/goog/crypt/hash32_test.html : Passed
00:38:31 : closure/goog/crypt/hmac_test.html : Loading
00:38:32 : closure/goog/crypt/hmac_test.html : Passed
00:38:32 : closure/goog/crypt/md5_test.html : Loading
00:38:33 : closure/goog/crypt/md5_test.html : Passed
00:38:33 : .../closure/goog/crypt/pbkdf2_test.html : Loading
00:38:33 : .../closure/goog/crypt/pbkdf2_test.html : Passed
00:38:33 : closure/goog/crypt/sha1_test.html : Loading
00:38:34 : closure/goog/crypt/sha1_test.html : Passed
00:38:34 : .../closure/goog/crypt/sha224_test.html : Loading
00:38:35 : .../closure/goog/crypt/sha224_test.html : Passed
00:38:35 : .../closure/goog/crypt/sha256_test.html : Loading
00:38:36 : .../closure/goog/crypt/sha256_test.html : Passed
00:38:36 : .../closure/goog/crypt/sha2_64bit_test.html : Loading
00:38:37 : .../closure/goog/crypt/sha2_64bit_test.html : Passed
00:38:37 : Finished
```

The currently shipped `all_tests.html` does not succeed on **all** tests, even on 2a79f2bc6b4.

### CODE REVIEW

**Comment by nanaze** [comment]
Note: @coruus is from the US. No concerns about accpeting crypto code.

File: closure/goog/crypt/sha12mc_test.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+goog.require('goog.crypt');
+goog.require('goog.crypt.Sha1');
+goog.require('goog.crypt.Sha224');
+goog.require('goog.crypt.Sha256');
+goog.require('goog.crypt.Sha384');
+goog.require('goog.crypt.Sha512');
+
+goog.require('goog.testing.jsunit');
+goog.require('goog.userAgent');
```
Comment (nanaze) [inline closure/goog/crypt/sha12mc_test.js]: not used

File: closure/goog/crypt/sha12mc_test.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
+/* Warning: These tests take about 8 minutes to run. */
+
+goog.provide('goog.crypt.ShaMcTest');
+goog.setTestOnly('goog.crypt.ShaMcTest');
+
+goog.require('goog.crypt');
```
Comment (nanaze) [inline closure/goog/crypt/sha12mc_test.js]: missing goog.array

**Comment by coruus** [comment]
The .travis.yml just compiles and lints Closure at the moment. goog.crypt.sha* lint clean.

File: .travis.yml [hunk 3]
```diff
@@ -5,10 +5,19 @@
 
 install:
   - cd ..
-  - git clone https://github.com/google/closure-compiler.git
+  - git clone --depth 1 https://github.com/google/closure-compiler.git
```
Comment (concavelenz) [inline .travis.yml]: I put together a separate PR for this. good idea:
https://github.com/google/closure-library/pull/538

File: .travis.yml [hunk 4]
```diff
@@ -5,10 +5,19 @@
...
   - cd closure-compiler
   - ant jar
   - cd ../closure-library
+  - sudo pip install svn+http://closure-linter.googlecode.com/svn/trunk#egg=gjslint
```
Comment (concavelenz) [inline .travis.yml]: The head travis script now runs the linter (from github)

File: .travis.yml [hunk 5]
```diff
@@ -5,10 +5,19 @@
...
-  - java -jar ../closure-compiler/build/compiler.jar -O ADVANCED --js='**.js' --js='!**_test.js' --js='!**_perf.js' --js_output_file=$(mktemp)
\ No newline at end of file
+  - java -jar ../closure-compiler/build/compiler.jar -O ADVANCED --js='**.js' --js='!**_test.js' --js='!**_perf.js' --js_output_file=$(mktemp)
+
+after_script:
+  - gjslint --strict --jslint_error=all --closurized_namespaces=goog --limited_doc_files='_test.js,datetimepatternsext.js' -r closure/goog
+
+env:
+  global:
+    - secure: chHRUDtNv3oSmOwrEu0SH/ML4SygVO5M0qwu7oUXj2CcHECTCbKuIf2R2jjyZqlJKn/efw+sdKhf0UTXXfmDIZuA53M+arGdrLGNewbeNVha1MY0hDVPOJKWo7O/miZj0hhFIO/qNhrvD/1uWD0y+jcXcWlqHOPWkwO3fITgzQY=
```
Comment (concavelenz) [inline .travis.yml]: Why is this necessary?

Comment (coruus) [inline .travis.yml]: This was for an integration with Sauce that I never bothered writing, IIRC. Inessential.

**Comment by coruus** [comment]
Any updates on this?

File: closure/goog/crypt/sha1.js [hunk 6]
```diff
@@ -184,19 +198,19 @@ goog.crypt.Sha1.prototype.compress_ = function(buf, opt_offset) {
...
     b = a;
     a = t;
   }
 
-  this.chain_[0] = (this.chain_[0] + a) & 0xffffffff;
-  this.chain_[1] = (this.chain_[1] + b) & 0xffffffff;
-  this.chain_[2] = (this.chain_[2] + c) & 0xffffffff;
-  this.chain_[3] = (this.chain_[3] + d) & 0xffffffff;
-  this.chain_[4] = (this.chain_[4] + e) & 0xffffffff;
+  this.chain_[0] = ((this.chain_[0] + a) & 0xffffffff) | 0;
```
Comment (shicks) [inline closure/goog/crypt/sha1.js]: As I understand the spec, the result of a bitwise & should already be a signed int32, so I'm puzzled as to how |0 will be any different (at least from my own experiments on Chrome).  Is this handling some edge case in another browser, or asm.js or something?

File: .travis.yml (hunk 5 above)
Comment (shicks) [inline .travis.yml]: The travis.yml changes should no longer be necessary - can you please remove them from this PR?

**Comment by joeltine** [comment]
@coruus are you still interested in this PR? Looks like there are several remaining comments that need to be addressed.

**Comment by joeltine** [comment]
Closing due to lack of response.
